src/v/kafka/client/direct_consumer/direct_consumer.cc

```diff
@@ -37,54 +40,90 @@ direct_consumer::direct_consumer(
 
 ss::future<fetches>
 direct_consumer::fetch_next(std::chrono::milliseconds timeout) {
+    // ss::lowres_clock's resolution is around the duration of task_quota and
+    // default task_quota is 500us
+    // adds a reasonable minimum timeout to calls to pop
+    static constexpr auto minimum_timeout = std::chrono::milliseconds{1};
     if (!_started) [[unlikely]] {
         throw std::runtime_error("Direct consumer is not started");
     }
     auto holder = _gate.hold();
+    auto deadline = ss::lowres_clock::now() + timeout;
+
     try {
-        auto maybe_response_to_filter = co_await _fetched_data_queue->pop(
-          timeout);
-        if (maybe_response_to_filter.has_error()) {
-            co_return maybe_response_to_filter;
-        }
+        // we'll keep attempting to pluck from the queue until the timeout is
+        // exhausted
+        while (ss::lowres_clock::now() < deadline) {
+            // either the remaining timeout or a small but reasonable minimum
+            // timeout
```

Comment:
may  I ask why this change is needed?  It seems like this timeout should be honored within the data_queue API if nothing is available to be fetched.